hphp/runtime/vm/jit/vasm-map-flags.cpp

```diff
@@ -0,0 +1,204 @@
+/*
+   +----------------------------------------------------------------------+
+   | HipHop for PHP                                                       |
+   +----------------------------------------------------------------------+
+   | Copyright (c) 2010-2016 Facebook, Inc. (http://www.facebook.com)     |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+*/
+
+#include "hphp/runtime/vm/jit/vasm.h"
+
+#include "hphp/runtime/vm/jit/abi.h"
+#include "hphp/runtime/vm/jit/containers.h"
+#include "hphp/runtime/vm/jit/vasm-gen.h"
+#include "hphp/runtime/vm/jit/vasm-instr.h"
+#include "hphp/runtime/vm/jit/vasm-print.h"
+#include "hphp/runtime/vm/jit/vasm-unit.h"
+#include "hphp/runtime/vm/jit/vasm-util.h"
+#include "hphp/runtime/vm/jit/vasm-visit.h"
+
+TRACE_SET_MOD(vasm);
+
+namespace HPHP { namespace jit {
+
+namespace {
+
+///////////////////////////////////////////////////////////////////////////////
+
+struct Env {
+  Vunit& unit;
+
+  // Map of VregSF to the inst that defs it
+  jit::hash_map<size_t, Vinstr*> sf_defs;
+
+  // Map of VregSF to list of insts that use it
+  jit::hash_map< size_t, jit::vector<Vinstr*> > sf_uses;
+};
+
+///////////////////////////////////////////////////////////////////////////////
+
+uint8_t getFlags(const Vinstr* inst) {
+  switch (inst->op) {
+#define O(name, imms, uses, defs) \
+    case Vinstr::name: { \
+      auto& i = inst->name##_; (void)i; \
+      imms \
+      break; \
+    }
+#define Inone
+#define fl fl; return i.fl
+#define I(x) (void)i.x;
```

Comment:
Oh, how about this—can you go through the opcode table and replace all the `I(cc)`s with `IC(cc)` and the `I(fl)`s with `IF(fl)`?  That way we can collect them explicitly using some `#define`s.  You'll just need to make sure that everywhere where we're currently defining `I()` and expecting it to hit all the immediates, we instead define `IF()` and `IC()` as well to be the same thing.